DNA

kovalevvv opened the pull request.

kovalevvv committed:
FINAL

slebedev commented on DNA/DNA/Cell+mutate.m:
```diff
@@ -0,0 +1,39 @@
+@implementation Cell (mutate)
+
+-(void) mutate:(int)x {
+    int indexes[x];
```
х - процентный, а не количественный показатель по условию задачи.

kovalevvv replied:
Всего 100 элементов в массиве, соответственно 1 элемент == 1%
Ошибаюсь?

slebedev replied:
Да, пока у вас 100 элементов. А если задать 450 элементов?

Да дело даже не в количестве элементов, я подозреваю, что вы сейчас будете говорить что-то типа "в задании указано 100 элементов и точка!". Точно так же можно сказать, что в задании указано что х - процентный показатель, и точка! Соответственно, количество элементов для изменения ещё нужно вычислить, исходя из этого процента.

kovalevvv replied:
Поправил

kovalevvv committed:
SUPE FINAL

kovalevvv committed:
TURBO FINAL

kovalevvv committed:
XTRA FINAL

slebedev commented on DNA/DNA/Cell+mutate.m:
```diff
@@ -0,0 +1,45 @@
+    if(p > 100 || p < 1) {
+        NSLog(@"Процентов должно быть больше 0 и меньше 100!");
+        exit(0);
+    }
```
Это не ошибка, просто поясню, что произойдёт в этом случае - после выполнения всех действий получится число с плавающей точкой, но так как это число будет присваиваться переменной типа int, дробная часть будет игнорироваться, будет использоваться целая часть. Так, например, при значении x=39.84 значение x будет 39. Если хотите использовать округление, то воспользуйтесь соответствующей функцией.

kovalevvv replied:
Поправил. Это РНР меня испортил. Спасибо!

kovalevvv committed:
COLLIDER FINAL

slebedev commented on DNA/DNA/Cell.m:
```diff
@@ -0,0 +1,48 @@
+
+-(void) printDNA {
+    for(int i = 0; i<100; i++)
+        NSLog(@"%@, %d", _DNA[i], i);
```
Вообще говоря, такой формат не удобен, да и если им пользоваться, то скорее делать по-привычному: 

`NSLog(@"%i: %@",i,_DNA[i]);`

Такой формат ведь гораздо логичнее: сначала показываем индекс, а затем то, что скрывается под этим индексом.

Но на деле удобнее, когда последовательность будет в одну строку, без указания каких-либо индексов. Вот, например, у класса NSMutableArray есть instance method (метод объекта класса) componentsJoinedByString:(NSString *), который вернёт все элементы в одну строку, соединённые с сепаратором, указанным в передаваемом параметре.

Тогда можно ограничиться лишь одной строкой:

`return [DNA componentsJoinedByString:@""];`

И ещё одно, что рассматривалось в сегодняшней лекции. Если ваш метод переделать в description, что возвращает строку, то тогда выводить вашу ДНК будет значительно проще:

`-(NSString *)description {
    return [DNA componentsJoinedByString:@""];
}`

А выводить можно так:

`Cell *myCell = [[Cell alloc] init];
NSLog(@"%@",myCell);`

kovalevvv replied:
спасибо!

kovalevvv committed:
HUBBLE FINAL

Bishop commented on DNA/DNA/Cell.m:
```diff
@@ -0,0 +1,46 @@
+
+-(id) init {
+    if(self = [super init]) {
+        NSArray *cDNA = [[NSArray alloc] initWithObjects:@"A",@"T",@"G",@"C", nil];
```
Эта строка у вас встречается в двух местах. Может логично вынести ее в метод.

Bishop commented on DNA/DNA/Cell+mutate.m:
```diff
@@ -0,0 +1,45 @@
+    NSArray *cDNA = [[NSArray alloc] initWithObjects:@"A",@"T",@"G",@"C", nil];
+    for(int i = 0; i<x; i++) {
+        
+        while(true) {
```
Очень неэффективный алгоритм. Итераций будет на несколько порядков больше, чем x. В худшем случае цикл действительно будет бесконечным.

Bishop commented on DNA/DNA/Cell+mutate.m:
```diff
@@ -0,0 +1,45 @@
+    }
+    
+    for(int i = 0; i<x; i++) {
+        [[self DNA] replaceObjectAtIndex:indexes[i] withObject:[cDNA objectAtIndex:arc4random()%4]];
```
Неточность. Например 
- DNA  содержит TGAAA...
- indexes[i] = 3
- [cDNA objectAtIndex:arc4random()%4] = A

Фактически замены не произойдет.

evgenpush replied:
Еще бы проверку добавить, чтобы символ на такой же не заменялся.

Bishop commented on DNA/DNA/Cell.h:
```diff
@@ -0,0 +1,19 @@
+
+@property NSMutableArray *DNA;
+
+-(void) printDNA;
```
Не вижу реализации этой функции

slebedev replied:
Он человеку и не нужен больше, он забыл его удалить из описания класса.

kovalevvv committed:
INFINITY NOT FINAL

kovalevvv committed:
ERROR NOT FINAL

kovalevvv commented on DNA/DNA/Cell+mutate.m:
```diff
@@ -0,0 +1,38 @@
+    
+    while(i<x) {
+        [[self DNA] replaceObjectAtIndex:rnd withObject:[[self cDNA] objectAtIndex:arc4random()%4]];
+        if([[[self DNA] objectAtIndex:rnd] isNotEqualTo:[copyDNA objectAtIndex:rnd]]) {
```
Кто-нибудь объяснит почему замены не происходит и условие, что А != А дает истину???

slebedev replied:
Почему вы так решили?

kovalevvv replied:
Выводим днк и тот же элемент в копии
https://www.dropbox.com/s/ydrizf4aozd209l/%D0%92%D1%8B%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5_006.png
это условие должно срабатывать когда элементы не равны (значит была замена) но и условие срабатывает и замены никакой не произошло :(

slebedev replied:
Просто вы сами не разобрались в том, что написали. =) Первой строкой цикла вы меняете элемент. Ну, если быть точным, то делаете предположение, что замена произошла. Затем вы это проверяете. То, что отработало тело "if" уже говорит о том, что замена произошла. Но затем вы что делаете? Генерируете новый индекс rnd! А потом по свежесгенерированному индексу выводите ячейку текущего объекта и ячейку копии. Естественно, они равны, так как вы не успели толком мутировать свою последовательность!

slebedev replied:
Надеюсь, я понятно объяснил? В самом начале мутации мы имеем два одинаковых массива! Вы берёте случайный индекс, заменяете по этому индексу элемент, и проверяете, не сходится ли он с тем, что был на этом месте в копии? Но выводите почему-то не до того, как сформировать новый индекс, а после, а ведь так как мы успели поменять только один элемент, по-любому значение по новому индексу будет сходным у двух массивов.

И ещё замечание: в задании указано, что элемент может меняться только один раз. Ваш алгоритм этого совсем не гарантирует.

kovalevvv replied:
Дада, я понял косяк с выводом, чтобы проверить. Мне стыдно :)
Понимаю, что может выпасть один и тот же индекс, что уменьшает количество замененных элементов. Решил, что надо сделать метод, который будет всегда возвращать только новый индекс элемента на замену. Столкнулся с тем, что в категории нельзя добавить к классу новых свойств. Пробовал категорию без имени

@interface Cell ()
@property NSMutableArray *idx;
@end

(в файле cell+mutate.h)

Компилятор это хавает, появляется свойство self.idx - хорошо, но как мне его наполнить значениями, чтобы использовать в качестве индексов? Этот массив хочет принимать только (id) ..

Я хочу так сделать, только потому, что есть метод [NSMutableArray containsObject:(id)] чтобы проверять существует ли уже такой индекс в массиве.

kovalevvv replied:
Просто я хочу добавить полноценную категорию, без изменения файлов головного класса Cell

slebedev replied:
Уважаемый @derats, прошу вас, пересмотрите лекцию 2. В ней вы увидите, что категория без названия - это расширение класса, и да, в нёй можно объявлять переменные класса. В категории можно объявлять и реализовывать только методы вашего объекта (или класса). Так же не помню точно, рассматривалось это в лекциях, или нет, но id - это "контейнер" для объекта. То есть, это, по сути, объект любого класса. Я подчёркиваю - объект класса! Переменная типа int - это не объект. Чтобы, например, проверить, есть ли в вашем массиве/множестве объект по заданному методу, нужно сделать примерно следующее (в случае переменной i):

```
    [DNA containsObject:[NSNumber numberWithInt:i]];
```

slebedev replied:
Грубо говоря, тип id - это обёртка для любого класса. Например:

```
    id tmp;  // переменная-"контейнер" типа id
    Cell *myCell = [[Cell alloc] init];
    tmp = myCell; // ок, tmp - это наш объект, можно сделать следующее:
    [tmp mutate:60]; // если выведете на экран после этого myCell, то увидите, что она изменилась
    tmp = [tmp DNA]; // а теперь tmp - типа NSMutableArray!
    NSLog(@"%@",[tmp objectAtIndex:10]); // выведет 10 элемент последовательности
```

То, что массив может принимать значения типа id означает, что его можно заполнить различными данными!
Вот, пример:

```
    NSNumber *number = [NSNumber numberWithInt:20];
    NSString *string = [NSString stringWithFormat:@"Тестовая строка"];
    NSArray *array = [NSArray arrayWithObjects:@"А это",@"тестовый",@"Массив", nil];
    NSMutableArray *newArray = [NSMutableArray array];
    [newArray addObject:number]; // добавляем в массив объект класса NSNumber
    [newArray addObject:string]; // и NSString
    [newArray addObject:newArray]; // и даже NSArray
```

Вполне спокойно, и всё отлично, у нас в массиве три объекта: число, строка, и ещё один массив из трёх строк.

kovalevvv replied:
Теперь я начал что-то понимать, спасибо!  Вы все очень понятно объясняете.

Но теперь у меня совсем ничего не работает, подозреваю дело в сеттере свойства, но как это исправить совершенно не понимаю.

Добавляю в класс новое свойство так
https://www.dropbox.com/s/bkheufmk7amia2l/%D0%92%D1%8B%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5_009.png

Пишу метод, но в итоге получается следующее:
https://www.dropbox.com/s/2b2rfacb0kw7i5s/%D0%92%D1%8B%D0%B4%D0%B5%D0%BB%D0%B5%D0%BD%D0%B8%D0%B5_010.png

Что-то я совсем сделал не так :(

В лекции немного по другому объявляют свойство, вот так
@interface Cell () {
 NSMutableArray *idx;
}
@end

В таком случае, похоже, почему-то к этому свойству IDE предлагает обращаться так self->idx 
Но и если делать так, то ошибка будет так же.

slebedev replied:
Зачем вам idx как переменная класса? Что в ней будет? Для чего она?

slebedev replied:
@derats сделав расширение класса, как это вы написали, вы объявили `@private` переменную класса Cell. У неё нет ни сеттеров, ни геттеров, поэтому в методах категории ничего не известно об этой переменной.

kovalevvv replied:
Это должен быть массив заменяемых индексов массива днк, чтобы метод next выдавал рандомный, но не повторяющийся новый индекс на замену. Где-то надо хранить эти значения.  

Я понимаю, что наверное можно определить эту переменную в классе cell и использовать в категории, но как то не верно когда переменные используемые только в категории объявляются в самом классе.

slebedev replied:
@derats ещё раз - методы категории работают с состоянием класса, состояние класса описывается в интерфейсе **класса**. Мне как данность не совсем понятно, зачем расширения придумали, вот это действительно может запутать. А категории работают логично - с теми переменными, что описаны в состоянии класса. Для генерации символа и символа на замену можно реализовать метод класса.

kovalevvv replied:
Мда, неразбериха пока что.
Я поправил, все работает, только не работает 100% мутация - приводит к бесконечному циклу. Не разобрался почему, с массивами все вроде в порядке, просто не может сгенерировать последний уникальный индекс.

kovalevvv committed:
CURIOSITY FINAL

Рабочий вариант, все заменяет и все проверки есть, но если методу
mutate  передать 100%, будет выполнятся бесконечно.

Думаю коммит HUBBLE FINAL более оптимальным был (всего 25% лишних
итераций)

Commit:
FINAL

Вот теперь уж точно финал))
